### PR TITLE
BXMSDOC-2577: dynamic tasks and process with the API

### PR DESCRIPTION
**Pull request by gsheldon**

### CODE REVIEW

File: docs/product-assembly_designing-and-building-cases/src/main/asciidoc/case-management-dynamic-user-task-API-proc.adoc [hunk 1]
```diff
@@ -0,0 +1,60 @@
+[id='case-management-dynamic-user-task-API-proc']
+= Creating a dynamic user task using the REST API
+
+You can create a dynamic user task during case run time using the REST API. To create a dynamic user task, the following information must be given:
+
+* task name
+* task description (optional, but recommended)
+* actors
+* groups
```
Comment (mswiderski) [inline docs/product-assembly_designing-and-building-cases/src/main/asciidoc/case-management-dynamic-user-task-API-proc.adoc]: either actor or groups or both can be given, it's not required to provide both users and groups

File: docs/product-assembly_designing-and-building-cases/src/main/asciidoc/case-management-dynamic-user-task-API-proc.adoc [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+* input data
+
+Use the following procedure to create a dynamic user task for the IT Orders sample project available in {CENTRAL} using the Swagger REST API tool. The same endpoint can be used for REST API without Swagger.
```
Comment (mswiderski) [inline docs/product-assembly_designing-and-building-cases/src/main/asciidoc/case-management-dynamic-user-task-API-proc.adoc]: I believe you can use Case management showcase app to add dynamic tasks

File: docs/product-assembly_designing-and-building-cases/src/main/asciidoc/case-management-adding-dynamic-tasks-using-API-proc.adoc [hunk 3]
```diff
@@ -0,0 +1,30 @@
+[id='case-management-adding-dynamic-tasks-using-API-proc']
+= Adding dynamic tasks to a case using the API
+
+Dynamic tasks and processes can be added to a case during run time to address unforeseen changes that can occur during the lifecycle of a case. Dynamic activities are not defined in the case definition and therefore, they cannot be signaled the way that a defined ad hoc task or process can.
```
Comment (emmurphy1) [inline docs/product-assembly_designing-and-building-cases/src/main/asciidoc/case-management-adding-dynamic-tasks-using-API-proc.adoc]: Make active voice. Remove comma after therefore.

File: docs/product-assembly_designing-and-building-cases/src/main/asciidoc/case-management-adding-dynamic-tasks-using-API-proc.adoc [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+The following dynamic activities can be added to a case:
```
Comment (emmurphy1) [inline docs/product-assembly_designing-and-building-cases/src/main/asciidoc/case-management-adding-dynamic-tasks-using-API-proc.adoc]: Active voice

File: docs/product-assembly_designing-and-building-cases/src/main/asciidoc/case-management-adding-dynamic-tasks-using-API-proc.adoc [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+* user tasks
```
Comment (emmurphy1) [inline docs/product-assembly_designing-and-building-cases/src/main/asciidoc/case-management-adding-dynamic-tasks-using-API-proc.adoc]: Make initial cap in lists.

File: docs/product-assembly_designing-and-building-cases/src/main/asciidoc/case-management-adding-dynamic-tasks-using-API-proc.adoc [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+* service tasks (any type that is implemented as a work item)
+* reusable subprocesses 
+
+Dynamic user and service tasks are added to a case instance and immediately executed. Depending on the nature of a dynamic task, it might start and wait for completion (user task) or directly complete after execution (service task). For dynamic subprocesses, the process engine requires a KJAR containing the process definition for that dynamic process in order to locate the process by its ID and execute it. This subprocess belongs to the case and has access to all of the data in the case file.
```
Comment (emmurphy1) [inline docs/product-assembly_designing-and-building-cases/src/main/asciidoc/case-management-adding-dynamic-tasks-using-API-proc.adoc]: delete 'in order'

File: docs/product-assembly_designing-and-building-cases/src/main/asciidoc/case-management-adding-dynamic-tasks-using-API-proc.adoc [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+You can use the Swagger REST API application to create dynamic tasks and subprocesses.
+
+.Prerequisite 
+You are logged in to {CENTRAL} and case instance has been started using the Showcase application. For more information about using Showcase, see
```
Comment (emmurphy1) [inline docs/product-assembly_designing-and-building-cases/src/main/asciidoc/case-management-adding-dynamic-tasks-using-API-proc.adoc]: and a case instance?

File: docs/product-assembly_designing-and-building-cases/src/main/asciidoc/case-management-adding-dynamic-tasks-using-API-proc.adoc [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+You can use the Swagger REST API application to create dynamic tasks and subprocesses.
+
+.Prerequisite 
+You are logged in to {CENTRAL} and case instance has been started using the Showcase application. For more information about using Showcase, see 
+ifeval::["{context}" == "case-management-design"]
+xref:case-management-showcase-application-con-case-management-design[Case management Showcase application].
+endif::[]
+
+.Procedure 
+. In a web browser, open the following URL: `http://localhost:8080/kie-server/docs`.
```
Comment (emmurphy1) [inline docs/product-assembly_designing-and-building-cases/src/main/asciidoc/case-management-adding-dynamic-tasks-using-API-proc.adoc]: URL on next line and no full stop.

Should :: be ->?

File: docs/product-assembly_designing-and-building-cases/src/main/asciidoc/case-management-adding-dynamic-tasks-using-API-proc.adoc [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+. Open the list of available endpoints under *Case instances :: Case Management*.
+. Locate the `POST` method endpoints for creating dynamic activities.
++
+image::swagger-case-management-dynamic.png[]
+. Click the REST endpoint required to create the dynamic task or process to open the documentation.
+. Click the *Try it out* button to provide the parameters and body required to create the dynamic activity.
```
Comment (emmurphy1) [inline docs/product-assembly_designing-and-building-cases/src/main/asciidoc/case-management-adding-dynamic-tasks-using-API-proc.adoc]: Click Try it out, (delete the button)

File: docs/product-assembly_designing-and-building-cases/src/main/asciidoc/case-management-dynamic-tasks-during-stages-proc.adoc [hunk 10]
```diff
@@ -1,2 +1,58 @@
-[id='']
-= TODO
\ No newline at end of file
+[id='case-management-dynamic-tasks-during-stages-proc']
+= Adding a dynamic task to a stages
+
+Dynamic tasks can be added to a case stage during run time using a REST API request. This is similar to adding a dynamic task to a case instance, but you must also define the `caseStageId` of the stage to which the task is added.
```
Comment (emmurphy1) [inline docs/product-assembly_designing-and-building-cases/src/main/asciidoc/case-management-dynamic-tasks-during-stages-proc.adoc:4]: Active

File: docs/product-assembly_designing-and-building-cases/src/main/asciidoc/case-management-dynamic-service-task-API-proc.adoc [hunk 11]
```diff
@@ -0,0 +1,58 @@
+[id='case-management-dynamic-service-task-API-proc']
+= Creating a dynamic service task using the REST API
+
+Service tasks are generally less complex than user tasks, though they might need more data to be provided to execute properly. Service tasks require following information:
```
Comment (emmurphy1) [inline docs/product-assembly_designing-and-building-cases/src/main/asciidoc/case-management-dynamic-service-task-API-proc.adoc]: although they...lose 'to be provided'
Server tasks require the following information

File: docs/product-assembly_designing-and-building-cases/src/main/asciidoc/case-management-dynamic-service-task-API-proc.adoc [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+
+* `name`: The name of the activity.
```
Comment (emmurphy1) [inline docs/product-assembly_designing-and-building-cases/src/main/asciidoc/case-management-dynamic-service-task-API-proc.adoc]: Inital caps in lists

Comment (emmurphy1) [inline docs/product-assembly_designing-and-building-cases/src/main/asciidoc/case-management-dynamic-service-task-API-proc.adoc]: I hate this, but IBM says to use this format for lists:
Data: the map... (no full stop unless at list one list item is a complete sentence

File: docs/product-assembly_designing-and-building-cases/src/main/asciidoc/case-management-dynamic-user-task-API-proc.adoc (hunk 2 above)
Comment (emmurphy1) [inline docs/product-assembly_designing-and-building-cases/src/main/asciidoc/case-management-dynamic-user-task-API-proc.adoc]: Consider reordering: 
 +Use the following procedure to use the Swagger REST API to create a dynamic user task  for the IT Orders sample project available in {CENTRAL} .

File: docs/product-assembly_designing-and-building-cases/src/main/asciidoc/case-management-dynamic-service-task-API-proc.adoc [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+
+. In a web browser, open the following URL:
++
+`http://localhost:8080/kie-server/docs`.
+. Open the list of available endpoints under *Case instances :: Case Management*.
+. Click click the following `POST` method endpoint to open the details: 
++
+`/server/containers/{id}/cases/instances/{caseId}/stages/{caseStageId}/tasks`
++
+. Click *Try it out* to complete the following parameters:
```
Comment (emmurphy1) [inline docs/product-assembly_designing-and-building-cases/src/main/asciidoc/case-management-dynamic-service-task-API-proc.adoc]: and complete..

File: docs/product-assembly_designing-and-building-cases/src/main/asciidoc/case-management-dynamic-tasks-during-stages-proc.adoc [hunk 14]
```diff
@@ -1,2 +1,58 @@
...
++
+id:: `itorders`
+caseId:: `IT-0000000001`
+caseStageId:: `Order delivery`
++
+The `caseStageId` is the name of the stage in the case definition where the dynamic task is to be created. 
++
+body:: This can be any dynamic or service task payload. See xref:case-management-dynamic-subprocess-API-proc[_Creating a dynamic subprocess using the REST API_] or xref:case-management-dynamic-service-task-API-proc[_Creating a dynamic service task using the REST API_] for examples.
+
+Once the dynamic task has been added to the stage, it must be completed in order for the stage to complete and for the case process to move on to the next case flow item.
```
Comment (emmurphy1) [inline docs/product-assembly_designing-and-building-cases/src/main/asciidoc/case-management-dynamic-tasks-during-stages-proc.adoc]: After the

File: docs/product-assembly_designing-and-building-cases/src/main/asciidoc/case-management-dynamic-service-task-API-proc.adoc [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+
+Service tasks are generally less complex than user tasks, though they might need more data to be provided to execute properly. Service tasks require following information:
+
+* `name`: The name of the activity.
+* `nodeType`: The type of node that will be used to find the work item handler.
+* `data`: The map of the data to properly deal with execution.
+
+You can create a dynamic service task during case run time with the same endpoint as a user task, but with a different body payload.
+
+Use the following procedure to create a dynamic service task for the IT Orders sample project available in {CENTRAL} using the Swagger REST API tool. The same endpoint can be used for REST API without Swagger.
```
Comment (emmurphy1) [inline docs/product-assembly_designing-and-building-cases/src/main/asciidoc/case-management-dynamic-service-task-API-proc.adoc]: Considering reordering AB4

File: docs/product-assembly_designing-and-building-cases/src/main/asciidoc/case-management-dynamic-subprocess-API-proc.adoc [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+.Procedure 
+. In a web browser, open the following URL:
++
+`http://localhost:8080/kie-server/docs`.
+. Open the list of available endpoints under *Case instances :: Case Management*.
+. Click click the following `POST` method endpoint to open the details: 
++
+`/server/containers/{id}/cases/instances/{caseId}/processes/{pId}`
++
+. Click *Try it out* to complete the following parameters:
```
Comment (emmurphy1) [inline docs/product-assembly_designing-and-building-cases/src/main/asciidoc/case-management-dynamic-subprocess-API-proc.adoc]: AB4

File: docs/product-assembly_designing-and-building-cases/src/main/asciidoc/case-management-dynamic-tasks-during-stages-proc.adoc [hunk 17]
```diff
@@ -1,2 +1,58 @@
...
++
+. In a web browser, open the following URL:
++
+`http://localhost:8080/kie-server/docs`.
+. Open the list of available endpoints under *Case instances :: Case Management*.
+. Click click the following `POST` method endpoint to open the details: 
++
+`/server/containers/{id}/cases/instances/{caseId}/stages/{caseStageId}/tasks`
++
+. Click *Try it out* to complete the following parameters:
```
Comment (emmurphy1) [inline docs/product-assembly_designing-and-building-cases/src/main/asciidoc/case-management-dynamic-tasks-during-stages-proc.adoc:48]: AB4

File: docs/product-assembly_designing-and-building-cases/src/main/asciidoc/case-management-dynamic-user-task-API-proc.adoc [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+
+. In a web browser, open the following URL:
++
+`http://localhost:8080/kie-server/docs`.
+. Open the list of available endpoints under *Case instances :: Case Management*.
+. Click click the following `POST` method endpoint to open the details: 
++
+`/server/containers/{id}/cases/instances/{caseId}/tasks`
++
+. Click *Try it out* to complete the following parameters:
```
Comment (emmurphy1) [inline docs/product-assembly_designing-and-building-cases/src/main/asciidoc/case-management-dynamic-user-task-API-proc.adoc:30]: AB4

File: docs/product-assembly_designing-and-building-cases/src/main/asciidoc/case-management-dynamic-service-task-API-proc.adoc [hunk 19]
```diff
@@ -0,0 +1,58 @@
...
+   "Operation" : "printMessage",
+   "ParameterType" : "java.lang.String"      
+  }, 
+ "nodeType" : "Service Task"
+}
+----
+. In the Swagger application, click *Execute* to create the dynamic task.
+
+
+In this example, a Java-based service is executed. It consists of an interface with the public class `org.jbpm.demo.itorders.services.ITOrderService`, public `printMessage` method with single `String` argument. Upon execution, the parameter value is passed to the method for execution.
```
Comment (emmurphy1) [inline docs/product-assembly_designing-and-building-cases/src/main/asciidoc/case-management-dynamic-service-task-API-proc.adoc]: method a with single

File: docs/product-assembly_designing-and-building-cases/src/main/asciidoc/case-management-dynamic-service-task-API-proc.adoc [hunk 20]
```diff
@@ -0,0 +1,58 @@
...
+
+Numbers and names, and other types of data given to create service tasks depends on the implementation of service task's handler. In the provided example, the `org.jbpm.process.workitem.bpmn2.ServiceTaskHandler` handler is used.
```
Comment (emmurphy1) [inline docs/product-assembly_designing-and-building-cases/src/main/asciidoc/case-management-dynamic-service-task-API-proc.adoc]: Delete comma
tasks depend... of a sevice task's...

File: docs/product-assembly_designing-and-building-cases/src/main/asciidoc/case-management-dynamic-service-task-API-proc.adoc [hunk 21]
```diff
@@ -0,0 +1,58 @@
...
+
+NOTE: For any custom service tasks, ensure the handler is registered in the deployment descriptor in *Work Item Handlers* section, where the name is same as the `nodeType` used for creating a dynamic service task.
```
Comment (emmurphy1) [inline docs/product-assembly_designing-and-building-cases/src/main/asciidoc/case-management-dynamic-service-task-API-proc.adoc]: in the deployment descriptor in the

File: docs/product-assembly_designing-and-building-cases/src/main/asciidoc/case-management-dynamic-subprocess-API-proc.adoc [hunk 22]
```diff
@@ -0,0 +1,46 @@
...
++
```
Comment (emmurphy1) [inline docs/product-assembly_designing-and-building-cases/src/main/asciidoc/case-management-dynamic-subprocess-API-proc.adoc]: Consider a better way to present this information. Perhaps a table with a Parameter, Value, and Description heading? Consider this for all similar cases.

File: docs/product-assembly_designing-and-building-cases/src/main/asciidoc/case-management-dynamic-user-task-API-proc.adoc [hunk 23]
```diff
@@ -0,0 +1,60 @@
...
+. In the Swagger application, click *Execute* to create the dynamic task.
+
+This procedure creates a new user task associated with case `IT-000000001`. The task is assigned to the person assigned to the `manager` case role. This task has two input variables:
+
+* `reason`
+* `caseFile_hwSpec`: defined as an expression to allow run time capturing of process or case data.
+
+Depending on the task, a form might be defined to provide a user-friendly UI for the task, which can be looked up by task name. In the IT Orders case, the `RequestManagerApproval` task includes the form `RequestManagerApproval-taskform.form` in its KJAR. 
+
+Once the task has been created, the task appears in the assignee`s *Task Inbox* in {CENTRAL}.
```
Comment (emmurphy1) [inline docs/product-assembly_designing-and-building-cases/src/main/asciidoc/case-management-dynamic-user-task-API-proc.adoc]: After the task.. make active if poss.

File: docs/product-assembly_designing-and-building-cases/src/main/asciidoc/case-management-dynamic-service-task-API-proc.adoc (hunk 12 above)
Comment (gsheldon) [reply docs/product-assembly_designing-and-building-cases/src/main/asciidoc/case-management-dynamic-service-task-API-proc.adoc]: This is kind of a definition list, for which IBM says "Capitalize the first word of each term or descriptive name, unless it usually would not be capitalized, for example, _num_circles_ or _prodname_ and capitalize the first word of each definition." 


The only solution I can think of here if it's really an issue is to format this like a terms list (ie, using :: to create a variable list), but I would still not be capitalizing these terms.

**Comment by gsheldon** [comment]
Thank you @mswiderski  and @emmurphy1 .

**Pull request merged**